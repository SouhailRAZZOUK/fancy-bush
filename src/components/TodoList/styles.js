import styled from "styled-components";

export const TodoListBlock = styled.ul({
  listStyle: "none",
  padding: 0
});
