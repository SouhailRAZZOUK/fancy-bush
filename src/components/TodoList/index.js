import { useCallback, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { deleteTodoItem, getTodoItems } from "../../redux/actions";
import { selectAllToDoItems } from "../../redux/selectors";
import TodoItem from "../TodoItem";
import { TodoListBlock } from "./styles";

const TodoList = ({ label }) => {
  const dispatch = useDispatch();
  const items = useSelector(selectAllToDoItems);
  const handleItemDelete = useCallback((targetId) => {
    dispatch(deleteTodoItem(targetId));
  }, [dispatch]);

  useEffect(() => {
    dispatch(getTodoItems());
  }, []);

  if (!items || items.length < 1) return <p>No Items</p>;

  return (
    <div>
      <h3>
        {label} ({items?.length})
      </h3>
      <TodoListBlock>
        {items?.map((item) => (
          <TodoItem key={item?.id} {...item} onItemDelete={handleItemDelete} />
        ))}
      </TodoListBlock>
    </div>
  );
};

export default TodoList;
