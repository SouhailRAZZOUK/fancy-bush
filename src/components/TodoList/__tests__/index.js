import { render } from "@testing-library/react";

import TodoList from "../";

const generateRandomDate = (start = new Date(), end = new Date()) => {
  if (!start || !end) return new Date();

  return new Date(
    start.getTime() + Math.random() * (end.getTime() - start.getTime())
  );
};
const TASKS_COUNT = 30;
const tasks = new Array(TASKS_COUNT)
  .fill({
    id: null,
    text: "Todo Item",
  })
  .map((item, index) => ({
    ...item,
    id: Math.random(),
    text: `${item.text} ${index + 1}`,
    createdAt: generateRandomDate(new Date(2022, 2, 1)),
    isDone: Math.random() * 10 > 5,
  }));

test("Todolist with many items", () => {
  const onItemDelete = jest.fn();
  const label = "Pending Items";

  const { container } = render(
    <TodoList items={tasks} onItemDelete={onItemDelete} label={label} />
  );

  // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
  const listItemsElements = container.querySelectorAll("li");
  expect(listItemsElements.length).toBe(TASKS_COUNT);

  // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
  const labelElement = container.querySelector("h3");
  expect(labelElement.textContent).toBe(`${label} (${TASKS_COUNT})`);

  //   expect(container).toBeFalsy()

  expect(container.innerHTML).toMatchSnapshot();
});

test("Todolist with no items", () => {
  const onItemDelete = jest.fn();

  const { container } = render(<TodoList onItemDelete={onItemDelete} />);

  // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
  const listItemsElements = container.querySelectorAll("li");
  expect(listItemsElements.length).toBe(0);

  // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
  const labelElement = container.querySelector("p");
  expect(labelElement.textContent).toBe("No Items");

  expect(container.innerHTML).toMatchSnapshot();
});
