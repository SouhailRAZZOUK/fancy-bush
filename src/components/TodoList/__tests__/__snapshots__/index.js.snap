// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todolist with many items 1`] = `"<div><h3>Pending Items (30)</h3><ul class=\\"sc-gsnTZi foEsZO\\"><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 1</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 2</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 3</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 4</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 5</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 6</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 7</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 8</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 9</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 10</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 11</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 12</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 13</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 14</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 15</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 16</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 17</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 18</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 19</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 20</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 21</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 22</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 23</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 24</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 25</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 26</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 27</li><li class=\\"sc-bczRLJ jcrRth\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 28</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 29</li><li class=\\"sc-bczRLJ eSbABr\\" style=\\"user-select: none; webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);\\">Todo Item 30</li></ul></div>"`;

exports[`Todolist with no items 1`] = `"<p>No Items</p>"`;
